Add `get` functionality to TaskGraph

richpsharp opened the pull request:
This PR fixes the following issues:

* Fixes #15 by adding a blocking, but timeoutable, `get` function to Tasks that return the return value of the Task's `func` after execution. 
* Helped #15 by adding `transient_run` flag to `add_task` that causes TaskGraph to avoid recording a completed `Task` so successive instatiations of the TaskGraph object will cause that function to reexecute even if it would have otherwise been precalculated.
* Fixes #17 by replacing a `time.sleep` with a `threading.Event` object that has a timeout on its `wait` method. During shutdown the event can be `set` to trigger an immediate wakeup.
* Fixes #16 by adding a `hash_target_files` flag to `add_task` that when False only checks the existence of the target files, not any other aspect of their status.
* Fixes #10 by removing one other uncessary lock.

Also updated history and relevant tests to cover these changes.

phargogh commented:
Hey @richpsharp I'll be very happy to take a look at this shortly, but I noticed that you have the PR set to be merged into `master` ... is this intended?  The last PR you put in, #14 , was merged into `release/0.9`, and this PR feels more like an enhancement than a bugfix.  Let me know what you think!  Wouldn't want any comments to become irrelevant if the PR destination were to change.

richpsharp commented:
I did not mean to do that James! I re-pointed the merge.

richpsharp commented on HISTORY.rst:
```diff
@@ -6,16 +6,44 @@ TaskGraph Release History
 Unreleased Changes
 ------------------
-* Updating primary repo url to Github.
+* Updating primary repository URL to GitHub.
```
I was running spell check on `HISTORY`!

richpsharp commented on HISTORY.rst:
```diff
@@ -6,16 +6,44 @@ TaskGraph Release History
+  ``transient_run`` is added to ``add_task`` that causes TaskGraph to avoid
+  recording a completed ``Task`` even if the execution hash would have been
+  identical to a previously completed run where the target artifacts still
+  existed.
```
Oh weird, I somehow thought I'd added new history but didn't even bother to check and somehow assumed I had forgotten. I like the second two jot points rather than the first so I removed the first in 3757937

richpsharp commented on taskgraph/Task.py:
```diff
@@ -421,8 +426,10 @@ def _task_executor(self):
+            hash_target_files=True, dependent_task_list=None,
+            ignore_directories=True, priority=0,
+            hash_algorithm='sizetimestamp', copy_duplicate_artifact=False,
+            transient_run=False):
```
YES! Originally I did what you suggested, but the problem is that we don't differentiate which files are hashed. When I say `exists` all I really care about is if the target path files exist. I can't think of a case where I want an `exists` on the target file but wouldn't want a regular hash on the input parameters. Hence the specific naming of `hash_target_files` defaulting to True. I hesitate to add an 'exists' hash algorithm to input files because I don't have a case where I've ever needed to do that. 

For what it's worth I've been using the `hash_target_files` syntax in some other projects and it feels initiative to me. 

So that's where that's at. I'm comfortable with this, but the next step would be to expand what `hash_algorithm` means and change it to some kind of per-argument parameter. I'd rather not do that until it's needed. What do you think?

phargogh replied:
Ah, thanks for clarifying the intent here!  I forgot that the `hash_algorithm` can apply to the input files as well.  Let's keep this as it is!

richpsharp commented on taskgraph/Task.py:
```diff
@@ -956,7 +1004,7 @@ def _call(self):
             return
         LOGGER.debug("not precalculated %s", self.task_name)
-        result_calculated = False
+        artifact_copied = False
```
Yes! Because now that we also recall return values, what's a result?

richpsharp commented on taskgraph/Task.py:
```diff
@@ -969,9 +1017,10 @@ def _call(self):
-                    result_target_path_stats = pickle.loads(
-                        database_result[0])
-                    LOGGER.debug('duplicate artifact db results: %s', result_target_path_stats)
+                    result_target_path_stats = pickle.loads(database_result[0])
```
Right now I think of a taskgraph database as really a temporary cache. Nice if it works but not critical if it doesn't. But if this turned into a centralized database that had a long life span we should start thinking about things like this. I'm not so worried that we'd need to guard against a corrupt pickle specifically (would the database be corrupt)? But we should have a mechanism for rollbacks, resets, and consistency checks. We don't have a use case for that now, but if we did we should think about it the same way any kind of mission critical database should be designed. I don't know how to do that. I'd probably read a book.

phargogh replied:
Yeah, I don't really know how to do that at well, and it feels like one of those things that's easy to conceive of and much harder to handle well.

The primary case I was worried about was the evolving standards for `pickle`.  For example, [python 3.4 introduced version 4, and 3.8 updated to version 5 of the pickle format](https://docs.python.org/3/library/pickle.html#data-stream-format), and I was worried that this update would break pickling.  At least for a simple test case of saving/loading a dict, I had no problems saving/loading between python 3.6 and 3.8 (which use different versions of the pickle spec).  I didn't try anything much more complicated (and there are some edge cases for `pickle` that can get hairy), but also we don't use `pickle` to save off super-complicated data structures and objects at this time.

So all this to say that what I was concerned about initially doesn't seem to be an issue, and yes, let's deal with how to handle this well when we start having a truly global state database we're maintaining in a cloud somewhere.

richpsharp commented on taskgraph/Task.py:
```diff
@@ -1002,27 +1051,31 @@ def _call(self):
+        if not self._hash_target_files:
+            target_hash_algorithm = 'exists'
+        else:
+            target_hash_algorithm = self._hash_algorithm
```
I get what you're saying, but I think this is still okay. I specifically want to differentiate between the "hash algorithm" and the "target file hash algorithm". Usually the "hash algorithm" is fine for all the files and if it's not it's only the case when the target files need to exist. 

In turn `_get_file_stats` shouldn't care if the "file" is input or output, just what it should do. So that's why I've got that intermediate step in this bit of code to differentiate what the `target_hash_algorithm` is on the one case when we only care if the files exist.

phargogh replied:
Yep, makes sense why this will be useful, and thanks for clarifying!

richpsharp commented on taskgraph/Task.py:
```diff
@@ -1037,19 +1090,23 @@ def _call(self):
         """Return true if _call need not be invoked.
 
+        If the task has been precalculated it will fetch the return result from
+        the previous run if possible.
```
Ha! Then the world ends. It always returns a value. `None` is a value. Changed in 04a8637

richpsharp commented on taskgraph/Task.py:
```diff
@@ -1162,20 +1228,55 @@ def join(self, timeout=None):
+            `hash_algorithm` is 'sizetimestamp' the task will require the
+            same base path files to determine equality. If it is a
+            `hashlib` algorithm only file contents will be considered. If this
+            value is 'exists' the value of the hash will be 'exists'.
```
Yes, because this is `_get_file_stats`. Here I do want to sometimes `exists` hash the files. It's up to whomever calls it. I'm really second guessing this stuff James, but it has not been casually implemented!

richpsharp commented:
Okay @phargogh, I see that you want more time to re-review it. I also took the liberty of addressing your comments/fixes that you've done so far. Also annoyed that I can't make threads on github's review page so I guess you know which comment I'm replying to. 

(and GitHub, yes I see that "quote reply", don't think I see you weaseling out of not implementing threads)

phargogh commented on taskgraph/Task.py:
```diff
@@ -1062,9 +1119,13 @@ def is_precalculated(self):
+        if not self._hash_target_files:
+            target_hash_algorithm = 'exists'
+        else:
+            target_hash_algorithm = self._hash_algorithm
```
👍 I'm now fine with it, as noted above!
